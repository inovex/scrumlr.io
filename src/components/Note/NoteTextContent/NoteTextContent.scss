@import "src/constants/style.scss";

.note-text-content-url {
  display: inline-block;

  &--truncate {
    max-width: 100%;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    vertical-align: top; /* align with the rest of the text */
    color: var(--accent-color);
  }
}
