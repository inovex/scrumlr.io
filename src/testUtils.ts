export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;