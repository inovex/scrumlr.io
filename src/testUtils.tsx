import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
