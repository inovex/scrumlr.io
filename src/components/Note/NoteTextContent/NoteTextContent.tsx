import Linkify from "linkify-react";
import classNames from "classnames";
import {addProtocol} from "utils/images";
import "./NoteTextContent.scss";

type NoteTextContentProps = {
  text: string;
  truncate?: boolean;
};

export const NoteTextContent = (props: NoteTextContentProps) => {
  // render a link, opening in a new tab.
  // this functionality could also be extended,
  // e.g. you could add wrappers for other scrumlr links etc.
  const renderLink = (content: {content: string}) => {
    const url = addProtocol(content.content);
    return (
      <a
        href={url}
        title={url}
        target="_blank"
        rel="noopener noreferrer"
        className={classNames("note-text-content-url", {"note-text-content-url--truncate": props.truncate})}
        onClick={(e) => e.stopPropagation()}
      >
        {content.content}
      </a>
    );
  };

  return (
    <Linkify
      options={{
        render: renderLink,
      }}
    >
      {props.text}
    </Linkify>
  );
};
