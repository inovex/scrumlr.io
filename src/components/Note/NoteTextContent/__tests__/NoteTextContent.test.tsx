import {NoteTextContent} from "../NoteTextContent";
import {render} from "testUtils";

describe("NoteTextContent", () => {
  it("should render correctly", () => {
    const testText = "hello https://scrumlr.io";
    const noteTextContent = render(<NoteTextContent text={testText} />);

    expect(noteTextContent.container).toMatchSnapshot();
  });

  it("should convert a plain URL to a clickable link", () => {
    const testText = "hello https://scrumlr.io";
    const noteTextContent = render(<NoteTextContent text={testText} />);

    expect(noteTextContent.container.getElementsByTagName("a")).toHaveLength(1);
  });

  it("should add a protocol for links without one", () => {
    const testText = "hello scrumlr.io";
    const noteTextContent = render(<NoteTextContent text={testText} />);
    const anchor = noteTextContent.container.getElementsByTagName("a")[0];

    expect(anchor).toHaveAttribute("href", expect.stringContaining("https://"));
  });

  it("should not add the truncated class if not defined", () => {
    const testText = "hello scrumlr.io";
    const noteTextContent = render(<NoteTextContent text={testText} />);
    const anchor = noteTextContent.container.getElementsByTagName("a")[0];

    expect(anchor).not.toHaveClass("note-text-content-url--truncate");
  });

  it("should add the truncated class if set to true", () => {
    const testText = "hello scrumlr.io";
    const noteTextContent = render(<NoteTextContent text={testText} truncate />);
    const anchor = noteTextContent.container.getElementsByTagName("a")[0];

    expect(anchor).toHaveClass("note-text-content-url--truncate");
  });
});
