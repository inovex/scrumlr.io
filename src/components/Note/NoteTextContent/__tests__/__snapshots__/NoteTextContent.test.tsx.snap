// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteTextContent should render correctly 1`] = `
<div>
  hello 
  <a
    class="note-text-content-url"
    href="https://scrumlr.io"
    rel="noopener noreferrer"
    target="_blank"
    title="https://scrumlr.io"
  >
    https://scrumlr.io
  </a>
</div>
`;
